src/runtime_src/core/pcie/windows/shim.h

```diff
@@ -19,6 +19,105 @@
 
 #include "xrt.h"
 #include "core/common/xrt_profiling.h"
+#include "boost/any.hpp"
+#include <boost/property_tree/ptree.hpp>
+#include "core/pcie/driver/windows/include/XoclUser_INTF.h"
+
+typedef enum {
+	pcie = 0,
+	rom,
+	icap,
+	xmc,
+	firewall,
+	dma,
+	dna
+}subdev;
+
+typedef enum {
+	VBNV = 0,
+	ddr_bank_size,
+	ddr_bank_count_max,
+	FPGA
+}rom_variable;
+
+typedef enum {
+	vendor = 0,
+	pcie_device,
+	subsystem_vendor,
+	subsystem_device,
+	link_speed,
+	link_width,
+	mig_calibration,
+	p2p_enable,
+	ready
+}pcie_variable;
+
+typedef enum {
+	clock_freqs = 0,
+	idcode
+}icap_variable;
+
+typedef enum {
+	detected_level = 0,
+	detected_status,
+	detected_time
+}firewall_variable;
+
+typedef enum {
+	version = 0,
+	serial_num,
+	max_power,
+	bmc_ver,
+	xmc_se98_temp0,
+	xmc_se98_temp1,
+	xmc_se98_temp2,
+	xmc_fpga_temp,
+	xmc_fan_temp,
+	fan_presence,
+	xmc_fan_rpm,
+	xmc_cage_temp0,
+	xmc_cage_temp1,
+	xmc_cage_temp2,
+	xmc_cage_temp3,
+	xmc_12v_pex_vol,
+	xmc_12v_pex_curr,
+	xmc_12v_aux_vol,
+	xmc_12v_aux_curr,
+	xmc_3v3_pex_vol,
+	xmc_3v3_aux_vol,
+	xmc_ddr_vpp_btm,
+	xmc_ddr_vpp_top,
+	xmc_sys_5v5,
+	xmc_1v2_top,
+	xmc_vcc1v2_btm,
+	xmc_1v8,
+	xmc_0v85,
+	xmc_mgt0v9avcc,
+	xmc_12v_sw,
+	xmc_mgtavtt,
+	xmc_vccint_vol,
+	xmc_vccint_curr,
+
+	xmc_3v3_pex_curr,
+	xmc_0v85_curr,
+	xmc_3v3_vcc_vol,
+	xmc_hbm_1v2_vol,
+	xmc_vpp2v5_vol,
+	xmc_vccint_bram_vol,
+	xmc_power
+}xmc_variable;
+
+void qr_rom_info(HANDLE handle, uint64_t variable, boost::any & _returnValue);
+void qr_pcie_info(HANDLE handle, uint64_t variable, boost::any & _returnValue);
+
+__declspec(dllexport) void queryDeviceWithQR(uint64_t _deviceID, uint64_t subdev, uint64_t variable, boost::any & _returnValue);
```

Comment:
This is not proper way of defining dllexport.  clients of dll should import, only the dll itself should export.
In fact I don't see why these functions must be exported given that they are only used locally within the libxrt_core.dll.

Reply:
queryDevcieWithQR() will be called from device_windows.cpp and this function implemented in shim layer. So, if we declare these functions without dllexport, we are seeing LINK errors.
Please suggest us how to fix this.

Reply:
You need to create a config.h file that sets up export and import macros for symbols that are exported out of libxrt_core.dll.    Take a look at runtime_src/core/common/config.h which serves the same purpose for the libxrt_coreutil.dll.  You could use XRT_CORE_SHIM_EXPORT/IMPORT in place of XRT_CORE_UTIL_EXPORT/IMPORT.
You include config.h in windows/shim.h and in shim.cpp you #define XRT_SHIM_SOURCE before including shim.h.